native/c/zusf.cpp

```diff
@@ -0,0 +1,190 @@
+    }
+    else
+    {
+      zusf->diag.e_msg_len = sprintf(zusf->diag.e_msg, "Path '%s' already exists! Mode: '%08o'", file.c_str(), file_stats.st_mode);
+    }
+    return RTNCD_FAILURE;
+  }
+
+  // TODO(zFernand0): Implement `mkdirp` by default
+  // TODO(zFernand0): `mkdirp` when creatnig a file in a directory that doesn't exist
+  if (createDir)
+  {
+    mkdir(file.c_str(), strtol(mode.c_str(), NULL, 8));
+    return RTNCD_SUCCESS;
+  }
+  else
+  {
+    ofstream out(file.c_str());
+    if (out.is_open())
+    {
+      out.close();
+      chmod(file.c_str(), strtol(mode.c_str(), NULL, 8));
+      return RTNCD_SUCCESS;
+    }
+  }
+
+  zusf->diag.e_msg_len = sprintf(zusf->diag.e_msg, "Could not create '%s'", file.c_str());
+  return RTNCD_FAILURE;
+}
+
+int zusf_list_uss_file_path(ZUSF *zusf, string file, string &response)
+{
+  struct stat file_stats;
+  if (stat(file.c_str(), &file_stats) == -1)
+  {
+    zusf->diag.e_msg_len = sprintf(zusf->diag.e_msg, "Path '%s' does not exist", file.c_str());
+    return RTNCD_FAILURE;
+  }
+
+  if (S_ISREG(file_stats.st_mode))
+  {
+    response.clear();
+    response += file.substr(file.find_last_of("/") + 1);
+    response.push_back('\n');
+    return RTNCD_SUCCESS;
+  }
+
+  if (!S_ISDIR(file_stats.st_mode))
+  {
+    zusf->diag.e_msg_len = sprintf(zusf->diag.e_msg, "Path '%s' is not a directory", file.c_str());
+    return RTNCD_FAILURE;
+  }
+
+  DIR *dir;
+  if ((dir = opendir(file.c_str())) == NULL)
+  {
+    zusf->diag.e_msg_len = sprintf(zusf->diag.e_msg, "Could not open directory '%s'", file.c_str());
+    return RTNCD_FAILURE;
+  }
+
+  struct dirent *entry;
+  response.clear();
+  while ((entry = readdir(dir)) != NULL)
+  {
+    // TODO(zFernand0): Skip hidden files
+    if ((strcmp(entry->d_name, ".") != 0) && (strcmp(entry->d_name, "..") != 0))
+    {
+      // TODO(zFernand0): Add option to list full file paths
+      // TODO(zFernand0): Add option to list file tags
+      response += entry->d_name;
+      response.push_back('\n');
+    }
+    // TODO(zFernand0): Sort in alphabetical order
+  }
+  closedir(dir);
+
+  return RTNCD_SUCCESS;
+}
+
+/**
+ * Reads data from a USS file.
+ *
+ * @param zusf pointer to a ZUSF object
+ * @param file name of the USS file
+ * @param response reference to a string where the read data will be stored
+ *
+ * @return RTNCD_SUCCESS on success, -1 on failure
+ */
+int zusf_read_from_uss_file(ZUSF *zusf, string file, string &response)
+{
+  ifstream in(file.c_str());
+  if (!in.is_open())
+  {
+    zusf->diag.e_msg_len = sprintf(zusf->diag.e_msg, "Could not open file '%s'", file.c_str());
+    return RTNCD_FAILURE;
+  }
+
+  in.seekg(0, ios::end);
+  size_t size = in.tellg();
+  in.seekg(0, ios::beg);
+
+  char *rawData = new char[size];
+  in.read(rawData, size);
+  in.seekg(0, ios::beg);
+
+  response.assign(rawData);
+  in.close();
+
+  return RTNCD_SUCCESS;
```

Comment:
I've fixed a memory leak for reading data sets and spool in #6, we'll probably want to fix it here too 😋 

```suggestion
  response.assign(rawData);
  in.close();
  
  delete[] rawData;

  return RTNCD_SUCCESS;
```

Reply:
great catch 🙏 

I'll check it out and implement the fixes here 🙏

Reply:
This seems to have been the only instance 🙏 
thanks for catching this 🙏